### PR TITLE
added burner on detection

### PR DESCRIPTION
**Pull request by melokda**
Changes:
- Thermal class uses contours to find burner temperatures
- Added burner on detection class to check for decreasing temperatures and pot on

### CODE REVIEW

File: SW/main/burner_on.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        if len(self.decreasing) == 12:
+            dx = np.diff(self.decreasing)
+            self.i = 0
+            self.decreasing = []
+            self.off = np.all(dx <= 0)
+            return np.all(dx <= 0)
+        else:
+            return None
+
+    def update(self, stove):
```
Comment (benhylak) [reply SW/main/burner_on.py]: (the rule of thumb is generally more than 2-3 times)

File: SW/main/Thermal.py [hunk 2]
```diff
@@ -24,46 +25,126 @@ def __init__(self):
...
+        bicubic_in = griddata(self.points, pixels_in, (self.grid_x, self.grid_y), method='cubic')
+
+        #input image has a cap on the maximum temperature
+        img = np.array(bicubic_in)
+
+        img.resize((32,32))
+
+        #split input and output images to represent each burner and get their temperatures
+
+        upper_left_in = img[0:16, 0:16]
```
Comment (benhylak) [inline SW/main/Thermal.py]: should refactor to have height and width variables (32 and 32)

File: SW/main/Thermal.py [hunk 3]
```diff
@@ -24,46 +25,126 @@ def __init__(self):
...
+        avg = 0.0
+
+        #turn input image into grayscale
+        blurred_in = cv2.GaussianBlur(img_in, (7, 7), 0)
+        im_in = np.stack((blurred_in,) * 3,-1)
+        im_in = im_in.astype(np.uint8)
+        bgr_in = cv2.cvtColor(im_in, cv2.COLOR_RGB2BGR)
+        gray_in = cv2.cvtColor(bgr_in, cv2.COLOR_BGR2GRAY)
+
+	#turn output image into grayscale
```
Comment (benhylak) [inline SW/main/Thermal.py:76]: having a hard time following -- add comments to explain why you're taking these steps

File: SW/main/burner_on.py (hunk 1 above)
Comment (benhylak) [inline SW/main/burner_on.py]: if you see lots of repeated code, make it a function!

File: SW/main/burner_on.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import Stove
+import numpy as np
+
+class BurnerOn:
```
Comment (benhylak) [inline SW/main/burner_on.py]: BurnerOnDetector

File: SW/main/burner_on.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            self.off = np.all(dx <= 0)
+            return np.all(dx <= 0)
+        else:
+            return None
+
+    def update(self, stove):
+        if (stove.upper_right.temp != 0):
+            self.decreasing_temp(stove.upper_right.temp)
+            upper_right_on = self.off
+            if stove.upper_right.pot_detected == False:
```
Comment (benhylak) [inline SW/main/burner_on.py]: i don't understand the purpose of this code.

Comment (benhylak) [reply SW/main/burner_on.py]: (checking if pot is detected here)

File: SW/main/burner_on.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        self.off = True
+        
+    def decreasing_temp(self, average):
+        while self.i < 360:
+            if (self.i % 30 == 0):
+                self.decreasing.append(average)
+            self.i += 1
+            break
+
+        if len(self.decreasing) == 12:
```
Comment (benhylak) [inline SW/main/burner_on.py]: lots of magic numbers here -- 360, 12 30, etc. 

Make them named variables so I have a chance of following along.

**Review by benhylak — CHANGES_REQUESTED** [review]

File: SW/main/burner_on.py (hunk 5 above)
Comment (melokda) [reply SW/main/burner_on.py]: Because we want to only check if burner is left on when pot is not on. Why not check it here?

**Comment by melokda** [comment]
Updates: 
- Added moving average class
- Edited the burner_on class to use moving average and update each burner correctly
- Edited Thermal class to return average of temperature over 10 seconds

Known issues:
- Taking the average of temperatures in thermal class reduces the effect of unexpected spikes but makes the change less noticeable in cases where temperature is decreasing very slowly
- The status of the burner fluctuates between on and off sometimes if there are unexpected spikes (such as another burner getting turned on). However, it doesn't take long for spikes to get moved out of moving array. So checking that the status of the burner has been on for a couple of minutes before sending a notification will help solve this problem.

File: SW/main/Thermal.py [hunk 7]
```diff
@@ -28,43 +29,66 @@ def __init__(self):
 
         self.points = [(math.floor(ix / 8), (ix % 8)) for ix in range(0, 64)]
         self.grid_x, self.grid_y = np.mgrid[0:7:32j, 0:7:32j]
+        
+        #height and width of processed image
+        self.height = 32
+        self.width = 32
+        
+        #minimum contour area
+        self.min_area = 20
```
Comment (benhylak) [inline SW/main/Thermal.py]: for constants, use all uppercase e.g. MIN_AREA -- makes it easier to read

File: SW/main/burner_on.py [hunk 8]
```diff
@@ -1,78 +1,45 @@
...
+from moving_average import MovingAverage
 
-class BurnerOn:
+class BurnerOnDetector:
     def __init__(self):
-        self.i = 0
-        self.decreasing = []
-        self.off = True
+        self.i = 0  #iteration time passed
+        self.sample_time = 10 #record a sample every 20 iterations
```
Comment (benhylak) [inline SW/main/burner_on.py]: change constants to uppercase e.g. SAMPLE_TIME

File: SW/main/burner_on.py [hunk 9]
```diff
@@ -1,78 +1,45 @@
...
+import random
+from moving_average import MovingAverage
 
-class BurnerOn:
+class BurnerOnDetector:
     def __init__(self):
-        self.i = 0
-        self.decreasing = []
-        self.off = True
+        self.i = 0  #iteration time passed
```
Comment (benhylak) [inline SW/main/burner_on.py]: i is usually used as an interation variable like

for(int i=0; i<3) etc. Rename

File: SW/main/burner_on.py [hunk 10]
```diff
@@ -1,78 +1,45 @@
...
-            if stove.upper_right.pot_detected == False:
-                if upper_right_on == False:
-                    stove.upper_right.on = True
-                else:
-                    stove.upper_right.on = False
-            else:
-                stove.upper_right.on = False
-        else:
-            stove.upper_right.on = False
+    def decreasing_temp(self, temp, moving_average):
```
Comment (benhylak) [inline SW/main/burner_on.py:19]: this should take a burner

detect_burner_on(self, burner)

**Review by benhylak — CHANGES_REQUESTED** [review]

File: SW/main/burner_on.py (hunk 10 above)
Comment (melokda) [reply SW/main/burner_on.py:19]: burner_on takes a burner, and decreasing_temp is called inside burner_on

Comment (benhylak) [reply SW/main/burner_on.py:19]: ah got it -- rename, then to detect_burner_on

File: SW/main/burner_on.py [hunk 11]
```diff
@@ -5,29 +5,29 @@
...
-        self.i = 0  #iteration time passed
-        self.sample_time = 10 #record a sample every 20 iterations
-        self.max_samples = 6 #the number of samples in the array
-        self.max_time = 200
-        self.moving_average_ur = MovingAverage(self.max_samples)
-        self.moving_average_ul = MovingAverage(self.max_samples)
-        self.moving_average_ll = MovingAverage(self.max_samples)
-        self.moving_average_lr = MovingAverage(self.max_samples)
-        self.min_temp = 20 #minimum temperature to consider burner on
+        self.main_iteration = 0  #iteration in the main file
```
Comment (benhylak) [inline SW/main/burner_on.py:8]: rename to loop_count

**Review by benhylak — APPROVED** [review]
squash/merge after changes are addressed

also -- in the future, someone told me once that commit comments should never be "addressed X's comments" -- should always be what you actually did.

**Pull request merged**